### ISSUE TITLE
daemonize after mount

### ISSUE THREAD

**Issue by comio** [rclone/rclone#2968]
#### What is your current rclone version (output from `rclone version`)?

```
rclone v1.46
- os/arch: linux/amd64
- go version: go1.11.5
```

#### What problem are you are trying to solve?

using mount + --daemon, rclone forks to early. Rclone should daemonize after the mount action in order to avoid sleeps after that.

#### How do you think rclone should be changed to solve that?

1) mount the mount point
2) Daemonize on success

Of course: Thanks for this impressive tools.

**Comment by ncw** [comment]
What are the bad consequences of the way it works at the moment?

Do you want to send in a fix?

**Comment by comio** [comment]
> What are the bad consequences of the way it works at the moment?
> 
> Do you want to send in a fix?

Hi,

my goal is to have a reliable mount.rclone wrapper. My actual solution is to use a sleep (or loop finding at least a file).
I think that a sort of solution is to call StartBackgroundMode() from mount.Mount() just after the mount(f, mountpoint) call, when we are sure that filesystem is really mounted. I'm not a go developer and I haven't any patch atm. I'm sorry for this.
I would underline that your creature works fine(TM) and this issue is just a suggestion from my side.

luigi

**Comment by comio** [comment]
Hi Again,

I failed my attempts to move StartBackGroundMode() after mount(f, mountpoint)... I suppose that go-routines are not happy if we daemonize. Very strange.
Another solution is to keep the daemonize where already is and use a sort of sync mechanism between the parent process and the child process (pipe?).  After the fork the parent waits on a pipe, the child, instead, will write on the pipe after the mount action.

**Comment by ncw** [comment]
> I failed my attempts to move StartBackGroundMode() after mount(f, mountpoint)... I suppose that go-routines are not happy if we daemonize. Very strange.

Yes go routines are not threads or processes they are [green threads](https://en.wikipedia.org/wiki/Green_threads).

You might want to check out [the help from the go-daemon library](https://github.com/sevlyar/go-daemon#how-it-works) rclone uses.

What it does is start a new process rather than call fork() which the go runtime is allergic to.

There doesn't seem to be a way to [pass data from child to parent yet](https://github.com/sevlyar/go-daemon/issues/19).

Perhaps what we could do is use the pidfile to monitor the child - get the parent to start the child, wait for N seconds and if the pidfile isn't there then tell the user the child has died and to check the log.

PS Looking at the code I think the pidfile handling is wrong - it appears to close it immediately after startBackGroundMode has returned.

**Comment by ncw** [comment]
I just noticed that [Reborn](https://godoc.org/github.com/sevlyar/go-daemon#Context.Reborn) returns the PID of the child process so instead of fiddling about with the pid file, we can just see if it is alive after N seconds which would be reasonably simple.

Maybe just calling [os.FindProcess](https://golang.org/pkg/os/#FindProcess) with the Pid from the [os.Process](https://golang.org/pkg/os/#Process) that [Reborn](https://godoc.org/github.com/sevlyar/go-daemon#Context.Reborn) returns.

**Comment by ivandeex** [comment]
This issue duplicates https://github.com/rclone/rclone/issues/992

**Comment by ivandeex** [comment]
The go-daemon issue https://github.com/sevlyar/go-daemon/issues/19 is 3 years stale without activity (as well as 14 other issues), while most of it fits within 5 small functions in https://github.com/sevlyar/go-daemon/blob/master/daemon_unix.go, and rclone uses 20% of its features (locks, signals, pidfile are unused)
I think we'd better admit that go-daemon solves a different use-case, drop it at least on linux, and re-implement it in a new rclone/cmd/mountlib/daemon_linux.go

**Comment by ivandeex** [comment]
I failed my attempts to move StartBackGroundMode() after mount(f, mountpoint)... I suppose that go-routines are not happy if we daemonize.

I'm not sure abou goroutines but I see other stuff that should run in child only and not in parent:
- [rc-server](https://github.com/rclone/rclone/blob/v1.52.3/cmd/cmd.go#L379) - if enabled, it will bind to port in parent and cause "cannot bind" in daemonized child
- [NewFsDir](https://github.com/rclone/rclone/blob/v1.52.3/cmd/mountlib/mount.go#L305) - it chains to backend-specific, unconstrained, generally fork-unsafe [Fs.NewFs](https://github.com/rclone/rclone/blob/v1.52.3/fs/cache/cache.go#L90)
- and more

This calls for refactoring, esp. [initConfig](https://github.com/rclone/rclone/blob/v1.52.3/cmd/cmd.go#L362) can be split into a function common to all commands to be used in `cobra.OnInitialize` and into another function to contain child-only steps. Most commands would inherit it from `rootCmd.PersistentPreRun`. Daemonizing commands would disable it by `PersistentPreRun: func(){return nil}` and then call from child... just an idea

> Another solution is to keep the daemonize where already is and use a sort of sync mechanism between the parent process and the child process (pipe?). 

Yes! The child could request passwords piped from parent's stdin and return status code.

> What it does is start a new process rather than call fork()

I'm not completely sure abou that. go-daemon uses [os.StartProcess](https://github.com/sevlyar/go-daemon/blob/master/daemon_unix.go#L96), which in turn ends up in...

**Comment by ncw** [comment]
First I'll just say that attempting to background go processes is hard. It is an unforunate legacy of the unix world in that in order to detatch a process you have to do the little fork / setsid / fork dance. The go runtime really does not like being forked unless you do it really early!

> > I failed my attempts to move StartBackGroundMode() after mount(f, mountpoint)... I suppose that go-routines are not happy if we daemonize.
> 
> @comio
> I'm not sure about goroutines but I see other stuff that should run in child only and not in parent:
> 
> * [rc-server](https://github.com/rclone/rclone/blob/v1.52.3/cmd/cmd.go#L379) - if enabled, it will bind to port in parent and cause "cannot bind" in daemonized child

That is annoying!

> * [NewFsDir](https://github.com/rclone/rclone/blob/v1.52.3/cmd/mountlib/mount.go#L305) - it chains to backend-specific, unconstrained, generally fork-unsafe [Fs.NewFs](https://github.com/rclone/rclone/blob/v1.52.3/fs/cache/cache.go#L90)

True

> * and more

Indeed
 
> This calls for refactoring, esp. [initConfig](https://github.com/rclone/rclone/blob/v1.52.3/cmd/cmd.go#L362) can be split into a function common to all commands to be used in `cobra.OnInitialize` and into another function to contain child-only steps. Most commands would inherit it from `rootCmd.PersistentPreRun`. Daemonizing commands would disable it by `PersistentPreRun: func(){return nil}` and then call from child... just an idea

I think that might work. It might be hard to keep it working though and I'd be worried the mechanism might be fragile.

> > Another solution is to keep the daemonize where already is and use a sort of sync mechanism between the parent process and the child process (pipe?).
> 
> Yes! The child could request passwords piped from parent's stdin and return status code.

This still has the deamonization problem as in rclone will start the --rc in the parent unless we do something about it.

> > What it does is start a new process rather than call fork()
> 
> @ncw
> [Nitpick mode on]: in the Linux world it's expressed as [fork() + immediate exec()](https://golang.org/src/syscall/exec_unix.go#L257) ;)

:-)

I think for rclone, starting `rclone mount` via systemd works so much better and has to be my preferred way of doing things.

If we can make `--daemon` work better without wrecking the internals of rclone then that would be good, but I have a feeling it is going to be like trying to untangle a ball of string getting the things to initialise in the parent or child.

**Comment by ivandeex** [comment]
> in order to detatch a process you have to do the little fork / setsid / fork dance.
> [snip]
> go runtime really does not like being forked unless you do it really early!
> [snip]
> It might be hard to keep it working though and I'd be worried the mechanism might be fragile.
> [snip]
> rclone will start the --rc in the parent unless we do something about it.
> [snip]
> If we can make `--daemon` work better without wrecking the internals of rclone then that would be good, but I have a feeling it is going to be like trying to untangle a ball of string getting the things to initialise in the parent or child.

@ncw
The key words to me are _entangled_ and _fragile_ i.e. we need a way to ensure reproduceability, stability and continuity.
I suggest amending the _mount_ test suite. It's already fragile: mount tests fail sporadically for me, though not frequently.
I think it'd be good to run it in github-actions jobs separate from the main suite (separate because this suite of tests is _expected to be unstable_): on ubuntu 16.04, 18.04, probably on freebsd (https://github.com/vmactions/freebsd-vm), probably with intent to extend to OSX later (does it have systemd yet?).
The new use cases should be added: rclone in fstab, as a systemd mount unit, as a systemd automount unit, plus edge cases (premature umount, systemctl stop/restart mnt-path.mount, automount idle timeout etc).
I didn't mention Windows since it lacks systemd and current mount test suite still suffices.
Let's start from mounting _:memory:_, after achieving some stability add _local:_, then one of major providers like _s3_.
Does this sound reasonable?

**Comment by ivandeex** [comment]
> it is going to be like trying to untangle a ball of string getting the things to initialise in the parent or child

daemonization will have 3 stages :
- _parent_ runs hand-picked subset of fork-safe operations from current `cmd.initConfig`, before it calls `os.StartProcess` (which does fork+exec on *nixes)
- _child setup_ takes place in the child which performs remaining unsafe operations from `initConfig` e.g. starts rc-server, calls `fs.NewFs`, sets up _vfs_, starts fuse mount and returns its status to parent
At this stage child has fully working stdin/stdout/stderr (either parent redirects child streams through iself or perhaps the child just sits attached to the parent's terminal session) and can log debug/error messages to stderr or request user input from stdin.
Parent just pipes child's streams and waits for status or timeout.
- _child detach_ - child replaces stdin/stdout/stderr by _/dev/null_, does _setsid_, probably does _setpgrp_ (detaches from process group) to ensure it will be later reaped by _pid 1_ when it exits or aborts, reopens log file descriptor and syslog connection if needed (because if any file descriptors remain from parent, the child ends up zombie later), and probably it detaches from original cgroup (if parent was run as a systemd unit) or proceeds to business as usual.
At this stage parent receives last status from child (or detects timeout), logs it and exits.

As a consequence, we have to:
- triage `initConfig` contents (it's rather small) between stages 1 and 2;
- think of switching child logger between file and redirected stdout(err);
- rethink config key;
- refactor all ReadLine, ReadPassword, fmt.Scanf and alikes in the core including _fs/config_ and in backends (where it reads  to use a common wrapper that can read terminal when available, read stdin(out) piped from parent at stage 2 or use universally configured helper executables(files) when detached or log loud errors when not configured;
- parent should issue warnings in advance if child's log file or stdin helper isn't configured.

**Comment by ivandeex** [comment]
As a side note, rclone has 10+ storage systems supported. Sooner or later one will start thinking about factoring backends into separate executables spawned by core and talking grpc/rest/whatever on top of rclone sdk. This issue can be viewed as an excersise or prelude for that.

**Comment by ivandeex** [comment]
Related issue: password prompt handling https://github.com/rclone/rclone/issues/524

**Comment by ivandeex** [comment]
Note: the _mount_ test suite should verify that a sample (probably dummy) shared plugin works as expected in daemon mode.

**Comment by ivandeex** [comment]
Handling of interactive user input via stdin/stdout is used by rclone core to get config key and by many backends to update such stuff as 2-factor authentication or storage pins.
Daemonization will have to provide a common rclone API and configuration options to get those from helper executables, external files, pipes etc
instead of stdin.
It's going to affect code in various backend, so I'd better submit a separate issue/PR.

**Comment by ivandeex** [comment]
Related to https://github.com/rclone/rclone/issues/3386 (password command)

**Comment by ivandeex** [comment]
Related to https://github.com/rclone/rclone/issues/4764 (integration hooks or events)

**Comment by ivandeex** [comment]
Related to https://github.com/rclone/rclone/issues/2346 (backend password from external command)

**Comment by ivandeex** [comment]
Issues related to _user input_ will be marked by the _terminal_ label. See https://github.com/rclone/rclone/labels/terminal

**Comment by ivandeex** [comment]
@comio 
Submitted patch #5593 should fix this issue, hopefully in 1.57
For now you can play with the latest `iva` build from https://github.com/ivandeex/rclone/releases

**Comment by ivandeex** [comment]
The PR should solve the original request of this ticket but a number of posts still stand:
- post https://github.com/rclone/rclone/issues/2968#issuecomment-716286668
- posts from https://github.com/rclone/rclone/issues/2968#issuecomment-716679470 to https://github.com/rclone/rclone/issues/2968#issuecomment-728309447

I will fork them into a separate ticket(s)
